### PR TITLE
Prevent Aborted (core dumped) in tf.nest.assert_same_structure for type mismatch in passed args

### LINKED ISSUE #42329: tf.nest.assert_same_structure crashes (abort) when some constraints are violated
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below):2.1.0
- Python version:3.7.6
- Bazel version (if compiling from source):N/A
- GCC/Compiler version (if compiling from source):N/A
- CUDA/cuDNN version:N/A
- GPU model and memory:N/A

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
`tf.nest.assert_same_structure` crashes (abort, systemerror) when some either `check_types` or `expand_composites` is 0d bool is violated.

**Describe the expected behavior**
Expect no crash

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

When `check_types` is boolean is violated
~~~python
import tensorflow as tf
import numpy as np
tf.nest.assert_same_structure(nest1=np.zeros((1)), nest2=tf.ones((1,1,1)), check_types=tf.ones((2)))
~~~
When `expand_composites` is boolean is violated
~~~python
import tensorflow as tf
import numpy as np
tf.nest.assert_same_structure(nest1=np.zeros((1)), nest2=tf.ones((1,1,1)), expand_composites=tf.ones((2)))
~~~

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
~~~python
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  SystemError: <class 'type'> returned a result with an error set

At:
  /root/miniconda3/lib/python3.7/site-packages/numpy/core/arrayprint.py(1388): _array_repr_implementation
  /root/miniconda3/lib/python3.7/site-packages/tensorflow/python/util/nest.py(397): assert_same_structure
  <stdin>(1): <module>

Aborted (core dumped)
~~~

### PR DESCRIPTION
**Pull request by yongtang**
This PR tries to address the issue raised in #42329 where
tf.nest.assert_same_structure will abort with core dump when
check_types is passed with a non-bool value.

The issue is related to pybind where type mismatch will
throw out an error and further cause 'pybind11::error_already_set' error.

This PR explicitly convert check_types to bool before passing to pybind,
and, in case check_types is not bool, an ValueError will be thrown out
gracefully by python itself.

This will be better than process abort (core dump).

This PR fixes #42329.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Review by mdanatg — DISMISSED** [review]

**Comment by rthadur** [comment]
@yongtang can you please check sanity build failures ?

**Comment by yongtang** [comment]
Thanks  @mdanatg @rthadur for the review. The PR has been updated with pylint failure fixed. Please take a look and let me know if there is any other issues.

**Review by mdanatg — APPROVED** [review]

File: tensorflow/python/util/nest_test.py [hunk 1]
```diff
@@ -1218,6 +1218,16 @@ def testListToTuple(self):
         expected,
     )
 
+  def testInvalidCheckTypes(self):
+    with self.assertRaises(ValueError):
```
Comment (mdanatg) [inline tensorflow/python/util/nest_test.py:1222]: This got flagged by internal tests: in graph mode, attempting to `bool()` cast a Tensor raises a specific error instead (`OperatorNotAllowedInGraphError`, a subclass of `TypeError`). I think the easiest way is to test that either ValueError os TypeError is raised.

**Review by mdanatg — CHANGES_REQUESTED** [review]

**Comment by yongtang** [comment]
Thanks @mdanatg for the help! The PR has been updated with captured exceptions expanded. Please take a look and see if the issue still exists.

**Review by mdanatg — APPROVED** [review]

**Pull request merged**